oxidation/libparsec/crates/core_fs/src/remote_loader.rs

```diff
@@ -0,0 +1,244 @@
+    remote_devices_manager: Arc<Mutex<RemoteDevicesManager>>,
+    realm_role_certificates_caches: Vec<RealmRoleCertificate>,
+}
+
+impl UserRemoteLoader {
+    pub fn new(
+        device: LocalDevice,
+        workspace_id: EntryID,
+        backend_cmds: AuthenticatedCmds,
+        remote_devices_manager: Arc<Mutex<RemoteDevicesManager>>,
+    ) -> Self {
+        Self {
+            device,
+            workspace_id,
+            backend_cmds,
+            remote_devices_manager,
+            realm_role_certificates_caches: vec![],
+        }
+    }
+
+    pub fn clear_realm_role_certificate_cache(&mut self) {
+        self.realm_role_certificates_caches.clear();
+    }
+
+    async fn load_realm_role_certificates_and_roles(
+        &mut self,
+        realm_id: Option<EntryID>,
+    ) -> FSResult<(Vec<RealmRoleCertificate>, HashMap<UserID, RealmRole>)> {
+        let realm_id = RealmID::from(*realm_id.unwrap_or(self.workspace_id));
+        let rep = self
+            .backend_cmds
+            .send(realm_get_role_certificates::Req { realm_id })
+            .await?;
+
+        match rep {
+            realm_get_role_certificates::Rep::Ok { certificates } => {
+                let mut unsecure_certifs = certificates
+                    .into_iter()
+                    .map(|uv| RealmRoleCertificate::unsecure_load(&uv).map(|rc| (rc, uv)))
+                    .collect::<DataResult<Vec<_>>>()
+                    .map_err(|e| FSError::InvalidRealmRoleCertificates(e.to_string()))?;
+
+                unsecure_certifs.sort_by(|a, b| a.0.timestamp.cmp(&b.0.timestamp));
+
+                let mut current_roles = HashMap::new();
+
+                for (unsecure_certif, raw_certif) in unsecure_certifs.iter() {
+                    let certif_author = match &unsecure_certif.author {
+                        CertificateSignerOwned::Root => {
+                            return Err(FSError::Custom(
+                                "Expected a certificate signed by a user".into(),
+                            ))
+                        }
+                        CertificateSignerOwned::User(device_id) => device_id,
+                    };
+
+                    let author = self
+                        .remote_devices_manager
+                        .lock()
+                        .await
+                        .get_device(certif_author, false)
+                        .await?;
+
+                    RealmRoleCertificate::verify_and_load(
+                        raw_certif,
+                        &author.verify_key,
+                        CertificateSignerRef::User(&author.device_id),
+                        None,
+                        None,
+                    )
+                    .map_err(|e| FSError::InvalidRealmRoleCertificates(e.to_string()))?;
+
+                    // Make sure author had the right to do this
+                    let needed_roles = match current_roles.get(&unsecure_certif.user_id) {
```

Comment:
Can you confirm that `RealmRoleCertificate::verity_and_load` goal is to verify if a certificate is valid ?

If yes, `RealmRoleCertificate::verify_and_load` should maybe return the trusted certificate.

That way we wouldn't use `unsecure_certif` later on the code (like for this match) but use the value return by `RealmRole...and_load` :thinking:

Reply:
Done